lago/providers/libvirt/cpu.py

```diff
@@ -0,0 +1,235 @@
+                self.cpu_model, vcpu_num=self.vcpu_num, host_cpu=self.host_cpu
+            )
+        else:
+            return self.generate_host_passthrough(self.vcpu_num)
+
+    def get_vcpu_xml(self, vcpu_num):
+        """
+        get_vcpu_xml
+
+        Args:
+            vcpu_num(int): number of virtual cpus
+
+        Returns:
+            lxml.etree.Element: vcpu XML node
+        """
+
+        return self.generate_vcpu(vcpu_num=self.vcpu_num)
+
+    def generate_host_passthrough(self, vcpu_num):
+        """
+        Generate host-passthrough XML cpu node
+
+        Args:
+            vcpu_num(int): number of virtual CPUs
+
+        Returns:
+            lxml.etree.Element: CPU XML node
+        """
+
+        cpu = et.Element('cpu', mode='host-passthrough')
+        cpu.append(self.generate_topology(vcpu_num))
+        return cpu
+
+    def generate_custom(self, cpu, vcpu_num, fill_topology):
+        """
+        Generate custom CPU model. This method attempts to convert the dict to
+        XML, as defined by ``xmltodict.unparse`` method.
+
+        Args:
+            cpu(dict): CPU spec
+            vcpu_num(int): number of virtual cpus
+            fill_topology(bool): if topology is not defined in ``cpu`` and
+                ``vcpu`` was not set, will add CPU topology to the generated
+                CPU.
+
+        Returns:
+            lxml.etree.Element: CPU XML node
+
+        Raises:
+            :exc:`~LagoInitException`: when failed to convert dict to XML
+        """
+
+        try:
+            cpu = utils.dict_to_xml({'cpu': cpu})
+        except:
+            # TO-DO: print an example here
+            raise LagoInitException('conversion of \'cpu\' to XML failed')
+
+        if not cpu.xpath('topology') and fill_topology:
+            cpu.append(self.generate_topology(vcpu_num))
+        return cpu
+
+    def generate_exact(self, model, vcpu_num, host_cpu):
+        """
+        Generate exact CPU model with nested virtualization CPU feature.
+
+        Args:
+            model(str): libvirt supported CPU model
+            vcpu_num(int): number of virtual cpus
+            host_cpu(lxml.etree.Element): the host CPU model
+
+        Returns:
+            lxml.etree.Element: CPU XML node
+        """
+
+        nested = {'Intel': 'vmx', 'AMD': 'svm'}
+        cpu = et.Element('cpu', match='exact')
+        et.SubElement(cpu, 'model').text = model
+        cpu.append(self.generate_topology(vcpu_num))
+
+        vendor = host_cpu.findtext('vendor')
+        if not nested.get(vendor):
+            LOGGER.debug(
+                'unknown vendor: {0}, did not configure nested '
+                'virtualization cpu flag on guest.'.format(vendor)
+            )
+            return cpu
+
+        flag = nested[vendor]
+        if self.host_cpu.find('feature/[@name="{0}"]'.format(flag)
+                              ) is not None:
+            cpu.append(self.generate_feature(name=flag))
```

Comment:
reminder to my self:
This fails when emulating AMD on Intel(as it will require svm, but check vmx exists). Need to also compare the cpu_model vendor given by the user.